File: day2.py
def part1():
    def tackle_line(x: list[int]) -> bool:
        if len(x) < 2:
            return True
        sign = x[1] > x[0]
        for i, y in enumerate(x[1:]):
            t = y - x[i]
            if ((t > 0) != sign) or abs(t) < 1 or abs(t) > 3:
                return False
        return True

    ans = 0
    with open("2024-2.txt", "r") as f:
        for line in f:
            x = line.split()
            x = list(map(int, x))
            ans += int(tackle_line(x))

    print(ans)


def part2():
    def tackle_line(x: list[int], try_again: bool = True) -> bool:
        if len(x) < 2:
            return True
        sign = x[1] > x[0]
        for i, y in enumerate(x[1:]):
            t = y - x[i]
            if ((t > 0) != sign) or abs(t) < 1 or abs(t) > 3:
                if try_again:
                    z = x.copy()
                    w = x.copy()
                    del x[i]
                    del z[i + 1]
                    del w[i - 1]
                    return (
                        tackle_line(x, False)
                        or tackle_line(z, False)
                        or tackle_line(w, False)
                    )
                else:
                    return False
        return True

    ans = 0

    with open("2024-2.txt", "r") as f:
        for line in f:
            x = list(map(int, line.split()))
            ans += int(tackle_line(x))

    print(ans)


part2()

File: src/bin/day2.rs
use std::{fs::File, io::read_to_string};

fn part1(line: &str) -> bool {
    let x: Vec<i32> = line.split(" ").map(|num| num.parse().unwrap()).collect();
    if x.len() < 2 {
        return true;
    }
    let sign = x[1] > x[0];
    for (i, num) in x[1..].iter().enumerate() {
        let diff = num - x[i];
        if ((diff > 0) != sign) || diff.abs() < 1 || diff.abs() > 3 {
            return false;
        }
    }
    true
}

fn part2(line: &str) -> bool {
    true
}

fn main() {
    let filename = "./2024-2.txt";
    let file = File::open(filename).unwrap();
    let mut count = 0;

    for line in read_to_string(file).unwrap().lines() {
        if part1(line) {
            count += 1;
        }
    }
    println!("{count}");

    let mut count2 = 0;
    for line in read_to_string(file).unwrap().lines() {
        if part2(line) {
            count2 += 1;
        }
    }
    println!("{count2}");
}
